tests/browser_context_test.go

```diff
@@ -688,3 +689,156 @@ func TestBrowserContextTimeout(t *testing.T) {
 		})
 	}
 }
+
+func TestBrowserContextWaitForEvent(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		name            string
+		event           string
+		optsOrPredicate *optsOrPredicate
+		wantErr         string
+	}{
+		{
+			name:  "successfully wait for page creation",
+			event: "page",
+		},
+		{
+			name:            "successfully wait for page creation with only predicate",
+			event:           "page",
+			optsOrPredicate: &optsOrPredicate{justPredicate: stringPtr("() => true;")},
+		},
+		{
+			name:            "successfully wait for page creation with predicate in option",
+			event:           "page",
+			optsOrPredicate: &optsOrPredicate{predicate: stringPtr("() => true;")},
+		},
+		{
+			name:            "successfully wait for page creation with predicate and timeout in option",
+			event:           "page",
+			optsOrPredicate: &optsOrPredicate{predicate: stringPtr("() => true;"), timeout: int64Ptr(1000)},
+		},
+		{
+			name:    "fails when event other than page passed in",
+			event:   "browser",
+			wantErr: "\"page\" is the only event that is supported, you passed in \"browser\"",
+		},
+		{
+			name:            "fails due to timeout",
+			event:           "page",
+			optsOrPredicate: &optsOrPredicate{predicate: stringPtr("() => false;"), timeout: int64Ptr(10)},
+			wantErr:         "waitForEvent timed out after 10ms",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			tb := newTestBrowser(t)
+
+			bc, err := tb.NewContext(nil)
+			require.NoError(t, err)
+
+			ctx, cancel := context.WithTimeout(tb.context(), 5*time.Second)
+			defer cancel()
+
+			var p1, p2 api.Page
+			err = tb.run(
+				ctx,
+				func() error {
+					op := optsOrPredicateToGojaValue(t, tb, tc.optsOrPredicate)
+					resp, err := bc.WaitForEvent(tc.event, op)
+					if resp != nil {
+						var ok bool
+						p1, ok = resp.(api.Page)
+						require.True(t, ok)
+					}
+					return err
+				},
+				func() error {
+					var err error
+					p2, err = bc.NewPage()
+					return err
+				},
+			)
+
+			if tc.wantErr == "" {
+				assert.NoError(t, err)
+				assert.Equal(t, p1.MainFrame().ID(), p2.MainFrame().ID())
+				return
+			}
+
+			assert.ErrorContains(t, err, tc.wantErr)
+		})
+	}
+}
+
+type optsOrPredicate struct {
```

Comment:
Can you help me understand what this type and functions do? Commenting inside the code would be great!

Reply:
Sure, i've added some comments in 8fd96c5507cd0d0f1d86f00b498123f08475d56d. Let me know if they help or not.

Reply:
Thanks, see: https://github.com/grafana/xk6-browser/pull/1042#discussion_r1334015121